nixos/lib/test-driver/test_driver/machine.py

```diff
@@ -720,6 +727,77 @@ def screenshot(self, filename: str) -> None:
             if ret.returncode != 0:
                 raise Exception("Cannot convert screenshot")
 
+    @contextmanager
+    def record_audio(
+        self,
+        filename: str,
+        allow_silence: bool = False,
+        device: str = "audio_device",
+        sample_rate: int = 44100,
+        bit_depth: int = 16,
+        channels: int = 2,
+    ) -> Iterator[None]:
+        filename = add_extension_if_simple_name(filename, "wav")
+        out_path = self.out_dir / filename
+
+        info_capture_regex = re.compile(
+            r"^\[(\d+)\]: Capturing audio\(\d+,\d+,\d+\) to (.*): (\d+) bytes$",
+        )
+
+        def capture_info() -> Optional[Tuple[int, int]]:
```

Comment:
reading this function feels like there is a lot of qemu-audio-capture specific domain logic and only a little bit of glue code that calls qemu-stuff and then throws it through the domain-specific code.

If you put the domain-specific part into a pure function, you would make it unit-testable and have test cases that at the same time document the interface that you expect from qemu. what do you think, does this make sense in your opinion?

Reply:
I feel like it is already pretty testable by calling it in weird conditions (capture already running at the same filename, invalid arguments, "manually" stopping the capture, etc.